File: go/algo/segment_tree/recursive.go
package segmenttree

type STR struct {
	n    int
	tree []int
	op   operation
}

func (st *STR) build(arr []int, treeIdx, l, r int) {
	if l == r {
		st.tree[treeIdx] = arr[l]
		return
	}

	mid := (l + r) >> 1
	st.build(arr, treeIdx<<1+1, l, mid)
	st.build(arr, treeIdx<<1+2, mid+1, r)

	st.tree[treeIdx] = st.op(st.tree[treeIdx<<1+1], st.tree[treeIdx<<1+2])
}

func NewSTR(nums []int, op operation) *STR {
	n := len(nums)
	tree := make([]int, n*4)
	st := &STR{
		n:    n,
		tree: tree,
		op:   op,
	}
	st.build(nums, 0, 0, n-1)
	return st
}

func (st *STR) query(treeIdx, segL, segR, l, r int) int {
	// inside range l...segL...segR...r
	if l <= segL && segR <= r {
		return st.tree[treeIdx]
	}

	mid := (segL + segR) >> 1
	if l > mid {
		// partial overlap {segL...[l...mid...segR}...r]
		// partial overlap [l....{segL...mid...r]...segR}
		// partial overlap {segL...[l....mid...r]...segR}
		return st.query(treeIdx<<1+2, mid+1, segR, l, r)
	} else if r <= mid {
		// partial overlap [l...{segL...r]...mid]...segR}
		// partial overlap {segL...[l...r]...mid]...segR}
		return st.query(treeIdx<<1+1, segL, mid, l, r)
	}

	// {segL...[l...mid...r]...segR}
	left := st.query(treeIdx<<1+1, segL, mid, l, mid)
	right := st.query(treeIdx<<1+2, mid+1, segR, mid+1, r)

	return st.op(left, right)
}

// Query performs operation query on [l; r] range
func (st *STR) Query(l, r int) int {
	return st.query(0, 0, st.n-1, l, r)
}

func (st *STR) update(treeIdx, l, r int, idx, val int) {
	if l == r {
		st.tree[treeIdx] = val
		return
	}

	mid := (l + r) >> 1
	if idx > mid {
		// l...mid...r...idx
		// l...mid...idx...r
		st.update(treeIdx<<1+2, mid+1, r, idx, val)
	} else {
		// l...idx...mid...r
		// idx...l...mid...r
		st.update(treeIdx<<1+1, l, mid, idx, val)
	}

	st.tree[treeIdx] = st.op(st.tree[treeIdx<<1+1], st.tree[treeIdx<<1+2])
}

// Update updates value of element w/ index idx
func (st *STR) Update(idx, value int) {
	st.update(0, 0, st.n-1, idx, value)
}

func (st *STR) Get(idx int) int {
	return 0
}

File: py/trees/segment_tree/recursive.py
"""
    https://codeforces.com/blog/entry/18051
    https://brestprog.by/topics/segmenttree/
    https://e-maxx.ru/algo/segment_tree
"""


# import operator
# ops = (
#     min,
#     max,
#     operator.add,
# )


class ST:
    __slots__ = 'op', 'tree', 'n'
    def __init__(self, nums: list[int], op):
        self.op = op

        self.tree = [0] * len(nums) * 4
        self.n = len(nums)
        self._build(nums, 0, 0, len(nums) - 1)

    def _build(self, nums: list[int], tree_idx: int, l: int, r: int):
        if l == r:
            self.tree[tree_idx] = nums[l]
            return

        mid = (l + r) >> 1
        self._build(nums, (tree_idx << 1) + 1, l, mid)
        self._build(nums, (tree_idx << 1) + 2, mid + 1, r)

        self.tree[tree_idx] = self.op(
            self.tree[(tree_idx<<1) + 1],
            self.tree[(tree_idx<<1) + 2]
        )

    def query(self, l: int, r: int):
        return self._query(0, 0, self.n-1, l, r)

    def _query(self, tree_idx: int, seg_l: int, seg_r: int, l: int, r: int):
        # segment is fully inside the range
        if l <= seg_l and seg_r <= r:
            return self.tree[tree_idx]

        mid = (seg_l + seg_r) >> 1
        if l > mid:
            return self._query((tree_idx<<1)+2, mid+1, seg_r, l, r)
        elif r <= mid:
            return self._query((tree_idx<<1)+1, seg_l, mid, l, r)

        return self.op(
            self._query((tree_idx<<1)+1, seg_l, mid, l, mid),
            self._query((tree_idx<<1)+2, mid+1, seg_r, mid+1, r)
        )

    def update(self, idx: int, value: int):
        ...
        # self._update

    def _update(self, tree_idx: int, l: int, r: int, idx: int, val: int):
        if l == r:
            self.tree[tree_idx] = val
            return

        mid = (l+r) >> 1
        if idx > mid:
            self.udpate((tree_idx<<1) + 2, mid+1, r, idx, val)
        else:
            self.update((tree_idx<<1)+1, l, mid, idx, val)

        self.tree[tree_idx] = self.op(self.tree[(tree_idx<<1)+1], self.tree[(tree_idx<<1)+2])
